### ISSUE TITLE
About CIFAR-10 pretrained weights

### ISSUE THREAD

**Issue by mountains-high** [fangwei123456/Parametric-Leaky-Integrate-and-Fire-Spiking-Neuron#5]
Hello~

I wanted to ask if it is possible to get only pre-trained CIFAR-10 weights?

I've been training for more than epoch=410 (several days) but current max_test_accuracy=0.8755, it hasn't increased since epoch=320.
Should we train to 1024 to get the same accuracy on paper?

Thank you for your consideration

```
speed per epoch 835.8156549930573
log_dir=logsd/CIFAR10_init_tau_2.0_use_plif_False_use_max_pool_True_T_8_detach_reset_True, max_test_accuracy=0.8755, train_times=1259375, epoch=403
Namespace(T=8, T_max=64, alpha_learnable=False, batch_size=16, channels=None, dataset_name='CIFAR10', detach_reset=True, device='cuda:0', init_tau=2.0, learning_rate=0.001, log_dir_prefix='logsd', max_epoch=410, normalization=None, number_layer=None, split_by=None, use_max_pool=True, use_plif=False)
train_clean.py -init_tau 2.0 -use_max_pool -use_plif -device cuda:0 -dataset_name CIFAR10 -log_dir_prefix logsd -T 8 -max_epoch 410 -detach_reset
test_accuracy 0.862
Writing....
Written.

```

**Comment by fangwei123456** [comment]
Hi, I just found the weights. But I am not sure it is the correct weights.

https://figshare.com/articles/dataset/Checkpoints_for_Parametric_Leaky_Integrate-and-Fire_Spiking_Neuron/19320206

Your accuracy is too low, and it should get 92+ accuracy after 400 epochs (see https://github.com/fangwei123456/Parametric-Leaky-Integrate-and-Fire-Spiking-Neuron/tree/main/logs/A/CIFAR10/CIFAR10_init_tau_2.0_use_plif_True_use_max_pool_True_T_8_detach_reset_True ).

**Comment by mountains-high** [comment]
Thank you.

**Comment by mountains-high** [comment]
Good day, @fangwei123456, 

I'm sorry for bringing this up again. I made only minor changes to make it work with the current version of the SpikingJelly and LIF neuron model only. It works fine, but the accuracy is low. 
```
log_dir=logsd/CIFAR10_init_tau_2.0_use_plif_False_use_max_pool_True_T_8_detach_reset_True, max_test_accuracy=0.8755, train_times=1737500, epoch=556

```
Could you help me solve this please? Am I doing something wrong? Thank you! 
```
#model.py
import torch
import torch.nn as nn
import torch.nn.functional as F

from spikingjelly.clock_driven import functional, layer, surrogate
from spikingjelly.clock_driven import neuron

from torchvision import transforms
import math
class StaticNetBase(nn.Module):
    def __init__(self, T, init_tau, use_max_pool, detach_reset):
        super().__init__()
        self.T = T
        self.init_tau = init_tau
        self.use_max_pool = use_max_pool
        self.detach_reset = detach_reset
        self.train_times = 0
        self.max_test_accuracy = 0
        self.epoch = 0
        self.static_conv = None
        self.conv = None
        self.fc = None
        self.boost = nn.AvgPool1d(10,10)

    def forward(self, x):
        x = self.static_conv(x)
        out_spikes_counter = self.boost(self.fc(self.conv(x)).unsqueeze(1)).squeeze(1)
        for t in range(1, self.T):
            out_spikes_counter += self.boost(self.fc(self.conv(x)).unsqueeze(1)).squeeze(1)

        return out_spikes_counter

class CIFAR10Net(StaticNetBase):
    def __init__(self, T, init_tau, use_max_pool, detach_reset):
        super().__init__(T, init_tau = init_tau, use_max_pool = use_max_pool, detach_reset = detach_reset)
        #1.
        self.static_conv = nn.Sequential(
            nn.Conv2d(3, 256, kernel_size = 3, padding = 1, bias = False),
            nn.BatchNorm2d(256)
        )
        #2.
        self.conv = nn.Sequential(
            #2.1
            neuron.LIFNode(tau = init_tau, surrogate_function = surrogate.ATan(), detach_reset = detach_reset),
            #2.2
            nn.Conv2d(256, 256, kernel_size=3, padding=1, bias=False),
            nn.BatchNorm2d(256),
            neuron.LIFNode(tau = init_tau, surrogate_function = surrogate.ATan(), detach_reset = detach_reset),
            #2.3
            nn.Conv2d(256, 256, kernel_size=3, padding=1, bias=False),
            nn.BatchNorm2d(256),
            neuron.LIFNode(tau = init_tau, surrogate_function = surrogate.ATan(), detach_reset = detach_reset),
            #2.4
            nn.MaxPool2d(2, 2),

            #2.5
            nn.Conv2d(256, 256, kernel_size=3, padding=1, bias=False),
            nn.BatchNorm2d(256),
            neuron.LIFNode(tau = init_tau, surrogate_function = surrogate.ATan(), detach_reset = detach_reset),
            #2.6
            nn.Conv2d(256, 256, kernel_size=3, padding=1, bias=False),
            nn.BatchNorm2d(256),
            neuron.LIFNode(tau = init_tau, surrogate_function = surrogate.ATan(), detach_reset = detach_reset),
            #2.7
            nn.Conv2d(256, 256, kernel_size=3, padding=1, bias=False),
            nn.BatchNorm2d(256),
            neuron.LIFNode(tau = init_tau, surrogate_function = surrogate.ATan(), detach_reset = detach_reset),
            #2.8
            nn.MaxPool2d(2, 2)
        )

        #3.
        self.fc = nn.Sequential(
            nn.Flatten(),
            layer.Dropout(0.5),
            nn.Linear(256 * 8 * 8, 128 * 4 * 4, bias=False),
            neuron.LIFNode(tau = init_tau, surrogate_function = surrogate.ATan(), detach_reset = detach_reset),
            nn.Linear(128 * 4 * 4, 100, bias=False),
            neuron.LIFNode(
                tau=init_tau, surrogate_function=surrogate.ATan(), detach_reset=detach_reset)
        )

```

```
#train.py
import torch
import torch.nn as nn
import torch.nn.functional as F
from spikingjelly.clock_driven import functional
import torchvision
from torch.utils.tensorboard import SummaryWriter
import numpy as np
import readline, os, argparse, time, sys

torch.backends.cudnn.benchmark = True
_seed_ = 2020
torch.manual_seed(_seed_)  # use torch.manual_seed() to seed the RNG for all devices (both CPU and CUDA)
torch.backends.cudnn.deterministic = True
torch.backends.cudnn.benchmark = False
np.random.seed(_seed_)
import models

if __name__ == '__main__':
    parser = argparse.ArgumentParser()
    #init_tau, batch_size, earning_rate, T_max, log_dir 
    parser.add_argument('-init_tau', type = float)
    parser.add_argument('-batch_size', type = int, default = 16)
    parser.add_argument('-learning_rate', type = float, default = 1e-3)
    parser.add_argument('-T_max', type = int, default = 64)
    parser.add_argument('-use_max_pool', action = 'store_true', default = True)
    parser.add_argument('-device', type = str)
    parser.add_argument('-dataset_name', type = str)
    parser.add_argument('-log_dir_prefix', type = str)
    parser.add_argument('-T', type = int)
    parser.add_argument('-channels', type = int)
    parser.add_argument('-number_layer', type = int)
    parser.add_argument('-split_by', type = str)
    parser.add_argument('-normalization', type = str)
    parser.add_argument('-max_epoch', type = int)
    parser.add_argument('-detach_reset', action = 'store_true', default = True)

    args = parser.parse_args()
    argv = ' '.join(sys.argv)

    print(args)
    init_tau = args.init_tau
    batch_size = args.batch_size
    learning_rate = args.learning_rate
    T_max = args.T_max
    use_max_pool = args.use_max_pool
    device = args.device
    dataset_name = args.dataset_name
    dataset_dir = '../data/' + dataset_name
    log_dir_prefix = args.log_dir_prefix
    T = args.T
    max_epoch = args.max_epoch
    detach_reset = args.detach_reset

    number_layer = args.number_layer
    channels = args.channels
    split_by = args.split_by
    normalization = args.normalization
    if normalization == 'None':
        normalization = None

    if dataset_name != 'MNIST' and dataset_name != 'FashionMNIST' and dataset_name != 'CIFAR10':
        dir_name = f'{dataset_name}_init_tau_{init_tau}_use_max_pool_{use_max_pool}_T_{T}_c_{channels}_n_{number_layer}_split_by_{split_by}_normalization_{normalization}_detach_reset_{detach_reset}'
    else:
        dir_name = f'{dataset_name}_init_tau_{init_tau}_use_max_pool_{use_max_pool}_T_{T}_detach_reset_{detach_reset}'

    log_dir = os.path.join(log_dir_prefix, dir_name)

    pt_dir = os.path.join(log_dir_prefix, 'pt_' + dir_name)
    print(log_dir, pt_dir)
    if not os.path.exists(pt_dir):
        os.mkdir(pt_dir)
    class_num = 10 #MNIST, FashionMNIST, CIFAR10
    if dataset_name == 'CIFAR10':
        transform_train, transform_test = models.get_transforms(dataset_name)

        train_data_loader = torch.utils.data.DataLoader(
                dataset  = torchvision.datasets.CIFAR10(
                    root = dataset_dir,
                    train = True,
                    transform = transform_train,
                    download  = True),
                batch_size = batch_size,
                shuffle = True,
                num_workers = 4,
                drop_last = False,
                pin_memory = True)
        test_data_loader = torch.utils.data.DataLoader(
                dataset = torchvision.datasets.CIFAR10(
                    root = dataset_dir,
                    train = False,
                    transform = transform_test,
                    download = True),
                batch_size = batch_size *16,
                shuffle = False,
                num_workers = 4,
                drop_last = False,
                pin_memory =True)



    check_point_path = os.path.join(pt_dir, 'check_point.pt')
    check_point_max_path = os.path.join(pt_dir, 'check_point_max.pt')

    net_max_path = os.path.join(pt_dir, 'net_max.pt')
    optimizer_max_path = os.path.join(pt_dir, 'optimizer_max.pt')
    sheduler_max_path = os.path.join(pt_dir, 'scheduler_max.pt')
    check_point = None

    if os.path.exists(check_point_path):
        check_point = torch.load(check_point_path, map_location = device)
        net = check_point['net']
        print(net.train_times, net.max_test_accuracy)
    else:
       if  dataset_name == 'CIFAR10':
            net = models.CIFAR10Net(T= T, init_tau = init_tau, use_max_pool = use_max_pool,
                    detach_reset = detach_reset).to(device)
    print(net)
    #Adam 
    optimizer=torch.optim.Adam(net.parameters(), lr = learning_rate)
    scheduler = torch.optim.lr_scheduler.CosineAnnealingLR(optimizer, T_max = T_max)

    if check_point is not None:
        optimizer.load_state_dict(check_point['optimizer'])
        scheduler.load_state_dict(check_point['scheduler'])
        log_data_list = check_point['log_data_list']
        del check_point
    else:
        log_data_list = []

    if log_data_list.__len__() > 0 and not os.path.exists(log_dir):
        rewrite_tb = True
    else:
        rewrite_tb = False
    writer = SummaryWriter(log_dir)
    if rewrite_tb:
        for i in range(log_data_list.__len__()):
            for item in log_data_list[i]:
                writer.add_scalar(item[0], item[1], item[2])

    if net.epoch != 0:
        net.epoch +=1
    ckpt_time = time.time()
    for net.epoch in range(net.epoch, max_epoch):
        start_time = time.time()

        log_data_list.append([])
        print(
                f'log_dir={log_dir}, max_test_accuracy={net.max_test_accuracy},train_times={net.train_times},epoch={net.epoch}')
        print(args)
        print(argv)

        net.train()
        for img, label in train_data_loader:
            img = img.to(device)
            label = label.to(device)

            optimizer.zero_grad()

            out_spikes_counter = net(img)
            out_spikes_counter_frequency = out_spikes_counter/net.T

            loss = F.mse_loss(out_spikes_counter_frequency, F.one_hot(label, class_num).float())

            loss.backward()
            optimizer.step()
            functional.reset_net(net) #resetting the state
            if net.train_times %256 == 0:
                accuracy = (out_spikes_counter_frequency.argmax(dim = 1) == label).float().mean().item()
                log_data_list[-1].append(('train_accuracy', accuracy, net.train_times))
                log_data_list[-1].append(('train_loss', loss.item(), net.train_times))
            net.train_times +=1
        scheduler.step()

        net.eval()
        with torch.no_grad():
            test_sum = 0
            correct_sum = 0
            for img, label in test_data_loader:
                img = img.to(device)
                label = label.to(device)
                out_spikes_counter = net(img)
                correct_sum += (out_spikes_counter.argmax(dim=1) == label).float().sum().item()
                test_sum += label.numel()
                functional.reset_net(net)
            test_accuracy = correct_sum / test_sum
            print('test_accuracy', test_accuracy)
            log_data_list[-1].append(('test_accuracy', test_accuracy, net.epoch))
            print('Writing ...')
            for item in log_data_list[-1]:
                writer.add_scalar(item[0], item[1], item[2])
            if net.max_test_accuracy <= test_accuracy:
                print('save model with test_accuracy =', test_accuracy)
                net.max_test_accuracy = test_accuracy
                torch.save({
                    'net':net,
                    'state': net.state_dict(),
                    'optimizer': optimizer.state_dict(),
                    'scheduler':scheduler.state_dict(),
                    'log_data_list': log_data_list
                    }, check_point_max_path)
            #Save the latest model
            torch.save({
                'net':net,
                'state':net.state_dict(),
                'optimizer':optimizer.state_dict(),
                'scheduler': scheduler.state_dict(),
                'log_data_list': log_data_list
                }, check_point_path)
            #if (time.time() - ckpt_time)>3600>6:
            #    #torch.save after every 6 hours
            #    torch.save({
            #        'net': net,
            #        'state': net.state_dict(),
            #        'optimizer': optimizer.state_dict(),
            #        'scheduler': scheduler.state_dict(),
            #        'log_data_list': log_data_list
            #    }, os.path.join(pt_dir, f'check_point_{net.epoch}.pt'))
            print('Written.')

            speed_per_epoch = time.time() - start_time
            print('speed per epoch', speed_per_epoch)
    writer.close()

```

**Comment by fangwei123456** [comment]
Coincidently, I just trained this network (without voting layer) with the new version of SJ. I only trained 128 epochs, and got acc1=91.53. I guess you will get higher accuracy if you train it with 1024 epochs.
![image](https://user-images.githubusercontent.com/38352109/167053162-4531a9ef-e1ff-4230-bfc0-da2e0e99a331.png)

`python main_plif_cf10.py --data-path /dev/shm/CIFAR10 -b 16 --opt adam --lr 0.001 --epochs 128 --amp --tb`
[cf10_plif.zip](https://github.com/fangwei123456/Parametric-Leaky-Integrate-and-Fire-Spiking-Neuron/files/8637040/cf10_plif.zip)

**Comment by mountains-high** [comment]
Thank you very much.
